### PR TITLE
Refactor HTTP Service Discovery

### LINKED ISSUE #6798: Refactor the HTTP service discovery configurations
### Component(s)

_No response_

### What is missing? Please describe.

Need to refactor our "tier 1" service discovery candidates.

### Describe alternatives you've considered.

_No response_

## Environment
Kubernetes Version:
Prometheus-Operator Version:

### PR DESCRIPTION
**Pull request by mviswanathsai**
## Description
This PR refactors the HTTP Service Discovery, adds new test cases for unit tests. 
PR also adds `OAuth2`, `FollowRedirects` and `EnableHTTP2` fields to the HTTP service discovery.

Generic API validation e2e tests are yet to be added.

Closes #6798 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

**Comment by mviswanathsai** [comment]
Not sure why `OAuth2`, `FollowRedirects` and `EnableHTTP2` were not added originally. I did not find any documentation stating that they are unsupported nor did I find any evidence for this in the code (hoping you could verify). Were these fields intentionally left out? was it a mistake?
CC: @slashpai @simonpasquier

File: pkg/prometheus/resource_selector.go [hunk 1]
```diff
@@ -1015,7 +1015,20 @@ func (rs *ResourceSelector) validateConsulSDConfigs(ctx context.Context, sc *mon
...
+	promVersion := operator.StringValOrDefault(rs.p.GetCommonPrometheusFields().Version, operator.DefaultPrometheusVersion)
+	version, err := semver.ParseTolerant(promVersion)
+	if err != nil {
+		return fmt.Errorf("failed to parse Prometheus version: %w", err)
+	}
+	if !version.GTE(semver.MustParse("2.28.0")) {
+		return fmt.Errorf("HTTP SD configuration is only supported for Prometheus version >= 2.28.0")
+	}
+
 	for i, config := range sc.Spec.HTTPSDConfigs {
```
Comment (simonpasquier) [inline pkg/prometheus/resource_selector.go:1027]: need to call rs.store.AddOAuth2() too?

**Comment by slashpai** [comment]
> Not sure why `OAuth2`, `FollowRedirects` and `EnableHTTP2` were not added originally. I did not find any documentation stating that they are unsupported nor did I find any evidence for this in the code (hoping you could verify). Were these fields intentionally left out? was it a mistake? CC: @slashpai @simonpasquier

httpsd was one of the first SD added to scrapeconfig IIRC it must have added required fields at that moment, thanks for adding rest

File: pkg/prometheus/resource_selector.go (hunk 1 above)
Comment (mviswanathsai) [reply pkg/prometheus/resource_selector.go:1027]: Ah! Thanks for pointing out, I will add it.

**Comment by mviswanathsai** [comment]
I will re-base once the EC2 PR is merged and add CRD validation tests for HTTP SD.

**Comment by mviswanathsai** [comment]
I will re-base as soon as we are happy with the changes and the PR for subtests is merged!

**Comment by simonpasquier** [comment]
Another merge conflict appeared :)

**Comment by mviswanathsai** [comment]
Conflicts resolved. PTAL @simonpasquier

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 2]
```diff
@@ -357,24 +357,35 @@ type HTTPSDConfig struct {
...
 	// BasicAuth information to authenticate against the target HTTP endpoint.
 	// More info: https://prometheus.io/docs/operating/configuration/#endpoints
+	// Cannot be set at the same time as `authorization`, or `oAuth2`.
 	// +optional
 	BasicAuth *v1.BasicAuth `json:"basicAuth,omitempty"`
 	// Authorization header configuration to authenticate against the target HTTP endpoint.
+	// Cannot be set at the same time as `oAuth2`, or `basicAuth`.
 	// +optional
 	Authorization *v1.SafeAuthorization `json:"authorization,omitempty"`
+	// Optional OAuth 2.0 configuration.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:375]: (nit)
```suggestion
	// Optional OAuth 2.0 configuration to authenticate against the target HTTP endpoint.
```

File: pkg/prometheus/resource_selector.go [hunk 3]
```diff
@@ -1042,7 +1042,20 @@ func (rs *ResourceSelector) validateConsulSDConfigs(ctx context.Context, sc *mon
 }
 
 func (rs *ResourceSelector) validateHTTPSDConfigs(ctx context.Context, sc *monitoringv1alpha1.ScrapeConfig) error {
+	promVersion := operator.StringValOrDefault(rs.p.GetCommonPrometheusFields().Version, operator.DefaultPrometheusVersion)
+	version, err := semver.ParseTolerant(promVersion)
+	if err != nil {
+		return fmt.Errorf("failed to parse Prometheus version: %w", err)
+	}
+	if !version.GTE(semver.MustParse("2.28.0")) {
```
Comment (simonpasquier) [inline pkg/prometheus/resource_selector.go:1050]:
```suggestion
	if rs.version.LT(semver.MustParse("2.28.0")) {
```

File: test/e2e/scrapeconfig_test.go [hunk 4]
```diff
@@ -540,6 +543,62 @@ func runScrapeConfigCRDValidation(t *testing.T, testCases []scrapeCRDTestCase) {
...
+			},
+		},
+		expectedError: true,
+	},
+	{
+		name: "Valid URL with http scheme",
+		scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
+			HTTPSDConfigs: []monitoringv1alpha1.HTTPSDConfig{
+				{
+					URL: "http://valid-url",
```
Comment (simonpasquier) [inline test/e2e/scrapeconfig_test.go:583]: (nit) a good habit is to use [reserved top-level domains](https://www.rfc-editor.org/rfc/rfc2606.html#page-2) for tests and documentation (the existing code base doesn't apply this recommendation though!). 
```suggestion
					URL: "http://valid.test",
```

**Review by simonpasquier — COMMENTED** [review]
LGTM, a few nits only.

File: test/e2e/scrapeconfig_test.go (hunk 4 above)
Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go:583]: Hmm... maybe we can just log an issue for that, it would make a good first-issue IMO.

**Comment by mviswanathsai** [comment]
Can't quite understand the unit-tests failing.
```
=== RUN   TestSelectScrapeConfigs/HTTP_SD_config_with_valid_proxy_settings
time=2024-08-30T11:26:59.002+05:30 level=WARN msg="skipping scrapeconfig" error="puppetDBSDConfigs: puppetDB SD configuration is only sup
ported for Prometheus version >= 2.31.0" scrapeconfig=test/test namespace=test prometheus=test
    resource_selector_test.go:3848: 
                Error Trace:    /home/mviswanathsai/operator/pkg/prometheus/resource_selector_test.go:3848
                Error:          "map[]" should have 1 item(s), but has 0
                Test:           TestSelectScrapeConfigs/HTTP_SD_config_with_valid_proxy_settings
=== RUN   TestSelectScrapeConfigs/HTTP_SD_config_with_valid_secret_ref
time=2024-08-30T11:26:59.003+05:30 level=WARN msg="skipping scrapeconfig" error="puppetDBSDConfigs: puppetDB SD configuration is only sup
ported for Prometheus version >= 2.31.0" scrapeconfig=test/test namespace=test prometheus=test
    resource_selector_test.go:3848: 
                Error Trace:    /home/mviswanathsai/operator/pkg/prometheus/resource_selector_test.go:3848
                Error:          "map[]" should have 1 item(s), but has 0
                Test:           TestSelectScrapeConfigs/HTTP_SD_config_with_valid_secret_ref
```

**Comment by mviswanathsai** [comment]
The test case seems right to me
```
		{
			scenario: "HTTP SD config with valid proxy settings",
			updateSpec: func(sc *monitoringv1alpha1.ScrapeConfigSpec) {
				sc.HTTPSDConfigs = []monitoringv1alpha1.HTTPSDConfig{
					{
						URL: "http://example.com",
						ProxyConfig: monitoringv1.ProxyConfig{
							ProxyURL:             ptr.To("http://no-proxy.com"),
							NoProxy:              ptr.To("0.0.0.0"),
							ProxyFromEnvironment: ptr.To(false),
							ProxyConnectHeader: map[string][]v1.SecretKeySelector{
								"header": {
									{
										LocalObjectReference: v1.LocalObjectReference{
											Name: "secret",
										},
										Key: "key1",
									},
								},
							},
						},
					},
				}
			},
			selected: true,
            promVersion: "2.29.0",
		},
```
Can't see why we get logs from `validatePuppetDBSDConfigs` in the error.

**Comment by mviswanathsai** [comment]
At the `resource_selector_test.go`, we could also do something similar  to what we did with the e2e tests for API validation and break up the large test in smaller subtests.

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**